sql/src/main/java/io/crate/action/sql/Session.java

```diff
@@ -376,49 +387,209 @@ public void execute(String portalName, int maxRows, ResultReceiver resultReceive
+                                throw new UnsupportedOperationException(
+                                    "Only write operations are allowed in Batch statements");
+                            }
+                            return singleExec(x.portal(), x.resultReceiver(), x.maxRows());
+                        }
+                    );
+                    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+                }
         }
     }
 
-    public void clearState() {
-        Portal portal = portals.remove(UNNAMED);
-        if (portal != null) {
-            portal.close();
+    private CompletableFuture<?> bulkExec(Map<Statement, List<DeferredExecution>> deferredExecutionsByStatement) {
+        var jobId = UUID.randomUUID();
+        var routingProvider = new RoutingProvider(Randomness.get().nextInt(), planner.getAwarenessAttributes());
+        var clusterState = executor.clusterService().state();
+        var txnCtx = new CoordinatorTxnCtx(sessionContext);
+        var plannerContext = new PlannerContext(
+            clusterState,
+            routingProvider,
+            jobId,
+            executor.functions(),
+            txnCtx,
+            0
+        );
+
+        var entry = deferredExecutionsByStatement.entrySet().iterator().next();
+        var statement = entry.getKey();
+        var toExec = entry.getValue();
+        var bulkArgs = Lists2.map(toExec, x -> (Row) new RowN(x.portal().params().toArray()));
+
+        Analysis analysis = analyzer.boundAnalyze(
+            statement, currentTxnCtx, new ParameterContext(Row.EMPTY, bulkArgs));
+
+        Plan plan;
+        try {
+            plan = planner.plan(analysis.analyzedStatement(), plannerContext);
+        } catch (Throwable t) {
+            PreparedStmt preparedStmt = toExec.get(0).portal().preparedStmt();
+            jobsLogs.logPreExecutionFailure(
+                jobId, preparedStmt.rawStatement(), SQLExceptions.messageOf(t), sessionContext.user());
+            throw t;
+        }
+        List<CompletableFuture<Long>> rowCounts = plan.executeBulk(
+            executor,
+            plannerContext,
+            bulkArgs,
+            SubQueryResults.EMPTY
+        );
+        CompletableFuture<Void> allRowCounts = CompletableFuture.allOf(rowCounts.toArray(new CompletableFuture[0]));
+        List<CompletableFuture<?>> resultReceiverFutures = Lists2.map(toExec, x -> x.resultReceiver().completionFuture());
+        CompletableFuture<Void> allResultReceivers = CompletableFuture.allOf(resultReceiverFutures.toArray(new CompletableFuture[0]));
+
+        return allRowCounts
+            .exceptionally(t -> null) // swallow exception - failures are set per item in emitResults
+            .thenAccept(ignored -> emitResults(jobId, jobsLogs, toExec, rowCounts))
+            .runAfterBoth(allResultReceivers, () -> {});
+    }
+
+    private static void emitResults(UUID jobId,
```

Comment:
would `emitRowCountResults` be a more descriptive name for this?

Reply:
Changed it to `emitRowCountsToResultReceivers`